ignite/cmd/plugin.go

```diff
@@ -71,6 +76,108 @@ func linkPluginCmds(rootCmd *cobra.Command, p *plugin.Plugin) {
 	}
 }
 
+func linkPluginHooks(rootCmd *cobra.Command, p *plugin.Plugin) {
+	if p.Error != nil {
+		return
+	}
+
+	for _, hook := range p.Interface.Hooks() {
+		linkPluginHook(rootCmd, p, hook)
+		if p.Error != nil {
+			return
+		}
+	}
+}
+
+func linkPluginHook(rootCmd *cobra.Command, p *plugin.Plugin, hook plugin.Hook) {
+	cmdPath := hook.PlaceHookOn
+
+	if !strings.HasPrefix(cmdPath, "ignite") {
+		// cmdPath must start with `ignite ` before comparison with
+		// cmd.CommandPath()
+		cmdPath = "ignite " + cmdPath
+	}
+
+	cmdPath = strings.TrimSpace(cmdPath)
+
+	cmd := findCommandByPath(rootCmd, cmdPath)
+
+	if cmd == nil {
+		p.Error = errors.Errorf("unable to find commandPath %q for plugin hook %q", cmdPath, hook.Name)
+		return
+	}
+
+	if !cmd.Runnable() {
+		p.Error = errors.Errorf("can't attach plugin hook %q to non executable command %s", hook.Name, hook.PlaceHookOn)
+		return
+	}
+
+	preRun := cmd.PreRunE
+	cmd.PreRunE = func(cmd *cobra.Command, args []string) error {
+		errors := make([]string, 0)
+		var err error
+		if preRun != nil {
+			err = preRun(cmd, args)
+			if err != nil {
+				// dont return the error, log it and let execution continue to `Run`
+				errors = append(errors, err.Error())
+			}
+		}
+
+		err = p.Interface.ExecuteHookPre(hook.Name, args)
+		if err != nil {
+			// dont return the error, log it and let execution continue to `Run`
+			errors = append(errors, err.Error())
+		}
+
+		time.Sleep(100 * time.Millisecond)
+		if len(errors) > 0 {
+			return fmt.Errorf("error while executing pre run hooks: %s", strings.Join(errors, ","))
+		}
+
+		return nil
+	}
+
+	postCmd := cmd.PostRunE
+	cmd.PostRunE = func(cmd *cobra.Command, args []string) error {
+		errors := make([]string, 0)
+
+		var err error
+		if preRun != nil {
+			err = postCmd(cmd, args)
+			if err != nil {
+				// dont return the error, log it and let execution continue to `Run`
+				errors = append(errors, err.Error())
```

Comment:
Same comment here, unless there's a good reason, I would quit immediately rather than continuing the execution. You can eventually execute the cleanup in a `defer` to ensure it will be executed even if there's an error.

Speaking of that, with this code there's always a chance for the cleanup phase to not being executed: when the `RunE` function itself returns an error. In that case the `PostRunE` are not executed by cobra. Maybe we should think of a different way to execute cleanups phase so they are executed whatever the result of the command. WDYT?

Reply:
Hmm yea, we could trigger it to run in the case of an error in 'RunE'. See my response to your other comment on letting the execution continue.